components/script/dom/readablestreambyobreader.rs

```diff
@@ -1,58 +1,202 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
-
 #![allow(dead_code)]
 
+use std::collections::VecDeque;
+use std::mem;
 use std::rc::Rc;
 
 use dom_struct::dom_struct;
 use js::gc::CustomAutoRooterGuard;
+use js::jsapi::Heap;
+use js::jsval::{JSVal, UndefinedValue};
 use js::rust::{HandleObject as SafeHandleObject, HandleValue as SafeHandleValue};
 use js::typedarray::ArrayBufferView;
 
+use super::readablestreamgenericreader::ReadableStreamGenericReader;
+use crate::dom::bindings::cell::DomRefCell;
 use crate::dom::bindings::codegen::Bindings::ReadableStreamBYOBReaderBinding::ReadableStreamBYOBReaderMethods;
 use crate::dom::bindings::error::Error;
 use crate::dom::bindings::import::module::Fallible;
-use crate::dom::bindings::reflector::{reflect_dom_object, DomObject, Reflector};
-use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::reflector::{reflect_dom_object_with_proto, DomObject, Reflector};
+use crate::dom::bindings::root::{DomRoot, MutNullableDom};
+use crate::dom::bindings::trace::RootedTraceableBox;
 use crate::dom::globalscope::GlobalScope;
 use crate::dom::promise::Promise;
 use crate::dom::readablestream::ReadableStream;
 use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
 
+/// <https://streams.spec.whatwg.org/#read-into-request>
+#[derive(JSTraceable)]
+pub enum ReadIntoRequest {
+    /// <https://streams.spec.whatwg.org/#byob-reader-read>
+    Read(Rc<Promise>),
+}
+
+impl ReadIntoRequest {
+    /// <https://streams.spec.whatwg.org/#read-into-request-chunk-steps>
+    pub fn chunk_steps(&self, _chunk: RootedTraceableBox<Heap<JSVal>>) {
+        todo!()
+    }
+
+    /// <https://streams.spec.whatwg.org/#read-into-request-close-steps>
+    pub fn close_steps(&self, _chunk: Option<RootedTraceableBox<Heap<JSVal>>>) {
+        todo!()
+    }
+
+    /// <https://streams.spec.whatwg.org/#read-into-request-error-steps>
+    pub(crate) fn error_steps(&self, _e: SafeHandleValue) {
+        todo!()
+    }
+}
+
 /// <https://streams.spec.whatwg.org/#readablestreambyobreader>
 #[dom_struct]
-pub(crate) struct ReadableStreamBYOBReader {
+pub struct ReadableStreamBYOBReader {
     reflector_: Reflector,
+
+    /// <https://streams.spec.whatwg.org/#readablestreamgenericreader-stream>
+    stream: MutNullableDom<ReadableStream>,
+
+    #[ignore_malloc_size_of = "Rc is hard"]
```

Comment:
There's no Rc. VecDeque is supported, so can we remove this ignore or push it deeper in the type hierarchy?